first stage of implementing LLVM code coverage

richkadel opened the pull request:
This PR replaces #70680 (WIP toward LLVM Code Coverage for Rust) since I am re-implementing the Rust LLVM code coverage feature in a different part of the compiler (in MIR pass(es) vs AST).

This PR updates rustc with `-Zinstrument-coverage` option that injects the llvm intrinsic `instrprof.increment()` for code generation.
    
This initial version only injects counters at the top of each function, and does not yet implement the required coverage map.

Upcoming PRs will add the coverage map, and add more counters and/or counter expressions for each conditional code branch.

Rust compiler MCP https://github.com/rust-lang/compiler-team/issues/278
Relevant issue: #34701 - Implement support for LLVMs code coverage instrumentation

***[I put together some development notes here, under a separate branch.](https://github.com/richkadel/rust/blob/cfa0b21d34ee64e4ebee226101bd2ef0c6757865/src/test/codegen/coverage-experiments/README-THIS-IS-TEMPORARY.md)***

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

richkadel commented:
> r? @davidtwco
> 
> (rust_highfive has picked a reviewer for you, use r? to override)

r? @tmandry @wesleywiser

richkadel commented:
> > (rust_highfive has picked a reviewer for you, use r? to override)
> 
> r? @tmandry @wesleywiser

richkadel commented:
@wesleywiser - PTAL. 

Note that Tyler and I also discussed the possibility of implementing the code generation with a custom Statement (rather than injecting a Call Terminator, as I did in this PR).

The Call Terminator allowed me to model the injected intrinsic the way that other intrinsics are handled (with a few important differences).

An alternative, suggested by Tyler, would be to define a new, custom StatementKind. Knowing that the `llvm.instrprof.increment` intrinsic will never panic, its BasicBlock does not branch, so a Statement would potentially avoid the need to inject a new BB.

My only hesitation is that a new StatementKind for this intrinsic becomes a very special case, different from all other intrinsics, and there may be hidden expectations and assumptions in the compiler that could add unexpected complexity.

But, I'm open to this alternative approach if there's a good reason to make that change.

Let us know your thoughts.

@tmandry - Please feel free to expand on or correct my explanation here. Thank you!

Aaron1011 commented on src/libcore/intrinsics.rs:
```diff
@@ -1946,6 +1946,22 @@ extern "rust-intrinsic" {
+/// `rustc_codegen_llvm::intrinsic::IntrinsicCallMethods::codegen_intrinsic_call()`.
+#[cfg(not(bootstrap))]
+#[cfg_attr(not(bootstrap), lang = "count_code_region")]
+fn count_code_region(_index: u32) {
```
Is it possible to make this a regular intrinsic? This would allow you to remove `InstanceDef::InjectedCode`.

richkadel replied:
I tried that at first but `InstanceDef::Intrinsic` implies a function signature with an `abi()` of `Abi::RustIntrinsic` or `Abi::PlatformIntrinsic`, which didn't fit with the implementation of the lang_item. Therefore it was more natural to create a new `InstanceDef` variant.

richkadel replied:
If your question is, why not make this a regular intrinsic in the `extern` block, in `libcore::intrinsics`, the short version of the long story is, I needed a DefId for the placeholder function injected in the MIR pass. The `lang_item` seems to be a good solution to that problem, but regardless, a DefId requires an actual definition, and `extern` functions aren't "defined" (that is, assigned a DefId) until they are called. Since the code for this call is injected very late in the process (MIR passes), I couldn't use the `extern` by itself, but I could get a DefId by defining the function with a body. This worked much better, but it cannot be handled exactly the same as a regular intrinsic.

Aaron1011 replied:
I don't think the ABI of the Rust function really matters here. You're manually constructing a call to the LLVM `instrprof_increment` intrinsic during codegen, which means that the Rust `count_code_region` function will not exist in the output IR at all.

I think the only ABI that actually matters here is the ABI of the underling LLVM intrinsic, which I don't think needs to be explicitly set.

Aaron1011 replied:
>  the short version of the long story is, I needed a DefId for the placeholder function injected in the MIR pass.

You can use `#[lang]` on an extern item, for example:

https://github.com/rust-lang/rust/blob/219380d840877de97b0e17674be7212ebc929d51/src/libcore/panicking.rs#L93-L96

richkadel replied:
Yes but only if it's called.

Aaron1011 replied:
All items in Rust end with a `DefId`, regardless of whether or not they are called. Did you run into an issue when trying to get the `DefId` of an extern item?

richkadel replied:
Yes, I did. As I understand it (and witnessed), when compiling a crate that does not call an extern, the extern is never pulled into the AST and therefore never lowered to HIR. There is not an "Item" for that extern function to resolve, so a DefId is never assigned. And so the DefId is never added to the lang_items table (if the lang item is defined on an extern fn).

For an extern that *is* called in the crate that's being compiled, it does get a DefId and that DefId is then associated with the lang_item, but for what I'm calling "InjectedCode", this is not going to be the case.

I did try many permutations with the intrinsic definition, as an extern, not an extern, with a function that calls the extern, etc. The final solution was the cleanest working solution I could come up with.

Aaron1011 replied:
> Yes, I did. As I understand it (and witnessed), when compiling a crate that does not call an extern, the extern is never pulled into the AST and therefore never lowered to HIR. 

The AST is created long before we determine which methods are actually called. Even if we could determine that a method was never called during AST construction, it would still need a `DefId` so that it can be be resolved by downstream crates.

I tried out your branch locally - I think I found the problem you ran into. The `lang_items` code emits some fairly confusing error messages.

When you place a `#[lang]` attribute on an `extern` item, it gets treated as a 'weak lang item' by `src/librustc_passes/weak_lang_items.rs`. However, the 'weak lang item' system is intended for lang items which eventually get linked to a symbol from some other crate - which is not how you're using `count_code_region`.

I would recommend adding a special case for your lang item in `Context::register`:

https://github.com/rust-lang/rust/blob/219380d840877de97b0e17674be7212ebc929d51/src/librustc_passes/weak_lang_items.rs#L72-L83

You'll want to skip all of the logic in that method for `count_code_region` - although it's a foreign (`extern`) item, it's not a "weak lang item" as understood by that code.

That should cause `count_code_region` to be treated like a normal lang item, allowing you to obtain its `DefId` and insert a call to it into the MIR.

richkadel replied:
Makes sense. I'll try it.

richkadel replied:
Nice! That did the trick. Thanks for the recommendation.

bjorn3 commented on src/libcore/intrinsics.rs:
```diff
@@ -1944,22 +1944,13 @@ extern "rust-intrinsic" {
+    /// option is enabled. The placeholder is replaced with `llvm.instrprof.increment` during code
+    /// generation.
+    #[cfg(not(bootstrap))]
+    #[cfg_attr(not(bootstrap), lang = "count_code_region")]
```
```suggestion
    #[lang = "count_code_region"]
```

wesleywiser commented on src/librustc_mir/transform/mod.rs:
```diff
@@ -265,6 +266,8 @@ fn mir_validated(
         &[&[
             // What we need to run borrowck etc.
             &promote_pass,
+            // FIXME(richkadel): is this the best place for the InstrumentCoverage pass?
```
I think we probably want this to run in the `post_borrowck_cleanup` group.

richkadel replied:
Done.

Note, when I moved the pass, I noticed that all of the `post_borrowck_cleanup` passes get called at least 3 times, when a macro like println!() is in the function body. 

It seems inefficient compared with where I had the pass before (which always only executed it once).

It still seems to work correctly (only codegen'ing the counter intrinsic once per function).

Is there a reason why these passes need to run multiple times? (And just making sure that reasoning also applies to the InstrumentCoverage pass.)

richkadel replied:
See comments below regarding moving this to `mir_validated()`.

wesleywiser commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -0,0 +1,100 @@
+    let new_block = START_BLOCK + body.basic_blocks().len();
+
+    let next_local = body.local_decls.len();
+    let new_temp = Local::new(next_local);
```
You might want to make use of [`rustc_mir::util::patch::MirPatch`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/util/patch/struct.MirPatch.html). It would probably simplify this code a fair bit.

richkadel replied:
Sounds good. I'll take another look at that. I referenced MirPatch when figuring out how to make that change and at first it was hard to tell how to insert a new start block. I think I have a better understanding now so I'll try to do that.

richkadel replied:
Done

Aaron1011 commented on src/librustc_codegen_llvm/intrinsic.rs:
```diff
@@ -136,6 +140,23 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
+
+                    let (fn_name, _len_val) = self.const_str(Symbol::intern(&caller_fn_path));
+                    let index = args[0].immediate();
+                    let hash = self.const_u64(1234);
```
The LLVM documentation is fairly vague about the `hash` value - is using `1234` intended to be permanent?

richkadel replied:
No, definitely not. Thanks for asking. This is just a stand-in for now. The `fn_name` will also be changed to a mangled function name.

tmandry replied:
Adding a FIXME would help avoid confusion.

richkadel replied:
Agreed. Will do.

richkadel replied:
Done

tmandry commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -0,0 +1,104 @@
+    val: u128,
+) -> Operand<'tcx> {
+    let param_env_and_ty = ty::ParamEnv::empty().and(ty);
+    let size = tcx
+        .layout_of(param_env_and_ty)
+        .unwrap_or_else(|e| panic!("could not compute layout for {:?}: {:?}", ty, e))
+        .size;
```
Since we'll be calling this function a lot and are already hardcoding the integer type, it probably makes sense to also hardcode `Size::from_bits()` instead of invoking a query.

richkadel replied:
Concur. I made slightly different changes that have some added benefits, but should have the same result.

tmandry commented on src/librustc_mir/transform/mod.rs:
```diff
@@ -296,6 +297,10 @@ fn run_optimization_passes<'tcx>(
         // `AddRetag` needs to run after `ElaborateDrops`. Otherwise it should run fairly late,
         // but before optimizations begin.
         &add_retag::AddRetag,
+        // If the `instrument-coverage` option is enabled, analyze the CFG, identify each
+        // conditional branch, construct a coverage map to be passed to LLVM, and inject counters
+        // where needed.
+        &instrument_coverage::InstrumentCoverage,
```
I'm now thinking that we should run this pass as early as possible before any optimization passes (i.e. before the first simplify_branches pass, _possibly_ after no_landing_pads). All the branches should look like they do in the source at that point. This will prevent some optimizations, but I think the priority should be accurate coverage info before all else.

richkadel replied:
@tmandry - As we discussed, I moved the `instrument_coverage` pass to the end of `mir_validate()` passes.

@wesleywiser - Tyler and I tried a few places including putting the coverage pass at the top of `post_borrowck_cleanup`. We agreed that it doesn't make sense to run `instrument_coverage` more than once per function. For reasons I don't yet understand, every `println!()` seems to cause `run_optimization_passes()` to run two additional times. (With 2 `println!()`calls in `main()` for example, my debug statement showed `instrument_coverage` ran 5 times on `main()` alone.)

Let me know if you disagree with the new location.

wesleywiser replied:
So I think what's actually happening is that when you add `println!()`s to the function body, some of the code added by that macro is eligible for [static promotion](https://github.com/rust-lang/rfcs/blob/master/text/1414-rvalue_static_promotion.md). Each promoted value gets its own `Body` which your code is then invoked for. Because these values were extracted out of the function, they use the function's `DefId` so when you call `debug!("instrumenting {:?}", src.def_id());`, you get the same `DefId` multiple times. If you log the full `src`, it should show that we're not running on the same `body` every time. 

You can see this by looking at the [Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ffb4b7013cea287855e78c81a383deb4). For a basic "hello world" application, if you click the "Show MIR" button (in the overflow menu button on the left hand side), you'll see that there is one body for `fn main()`, one body for a `&[std::fmt::ArgumentV1; 0]` which got promoted and one body for a `&[&str; 1]` which also got promoted. 

All that said, I don't have a problem with landing this as-is. We can always tweak this later and I think it's more important that we land something so we can start iterating on the coverage info generation.

tmandry commented:
@bors d+

r=me after comments

richkadel commented on src/librustc_mir/interpret/intrinsics.rs:
```diff
@@ -375,6 +375,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 );
                 self.copy_op(self.operand_index(args[0], index)?, dest)?;
             }
+            sym::count_code_region => (),
```
Noted. As discussed 1-on-1, I think you plan to file a Rust issue on this topic. Thanks!

tmandry replied:
#73156

oli-obk replied:
Please leave a FIXME comment here linking to the issue

tmandry commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -77,17 +79,24 @@ fn const_int_operand<'tcx>(
-        .layout_of(param_env_and_ty)
-        .unwrap_or_else(|e| panic!("could not compute layout for {:?}: {:?}", ty, e))
-        .size;
+    debug_assert!({
```
This might be redundant with MIR typeck, not sure. Can't hurt, though.

richkadel replied:
Yeah, I assume there are more checks, but this debug-only check makes it easier to pinpoint where the problem is introduced.

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 84d4424ff88db00736e940da9443c29ef5911101 has been approved by `tmandry`

<!-- @bors r=tmandry 84d4424ff88db00736e940da9443c29ef5911101 -->
<!-- homu: {"type":"Approved","sha":"84d4424ff88db00736e940da9443c29ef5911101","approver":"tmandry"} -->

richkadel commented:
Did not mean to close the PR. The GitHub app on Android has a Pull Request status window with a "Close" button. It looks like that button is how you close the window, but it closes the PR... Ach!

oli-obk commented:
Please add tests in `src/tests/mir-opt` that show various things that get annotated by the annotation pass.

You'll need to add `// compile-flags: -Zinstrument-coverage` and `// EMIT_MIR rustc.some_function_you_want_to_see_the_mir_of.InstrumentCoverage.diff` to the test file in order to get dumps.

oli-obk commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -0,0 +1,113 @@
+pub struct InstrumentCoverage;
+
+/**
+ * Inserts call to count_code_region() as a placeholder to be replaced during code generation with
```
Make this a doc-comment on the struct

oli-obk commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -0,0 +1,113 @@
+    let span = body.span.shrink_to_lo();
+
+    let count_code_region_fn =
+        function_handle(tcx, span, tcx.lang_items().count_code_region_fn().unwrap());
```
Please use `require_lang_item` instead of unwrapping the lang item (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.require_lang_item). This gives better diagnostics if the lang item is missing.

oli-obk commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -0,0 +1,113 @@
+    Operand::function_handle(tcx, fn_def_id, substs, span)
+}
+
+fn const_int_operand<'tcx>(
```
This could be an inherent method on `Operand`

Dylan-DPC commented:
fails in [rollup log](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32072&view=logs&j=9121965e-722b-5094-4a33-f061e41958d8&t=64b964f2-043e-5afc-0fc3-c41881c5cd34)

@bors r-

richkadel commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -57,6 +57,7 @@ pub fn instrument_coverage<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
             destination: Some((Place::from(temp), new_block)),
             cleanup: None,
             from_hir_call: false,
+            fn_span: span,
```
Fixed new merge error.

richkadel replied:
Tested locally, including the mir-opt test. This should fix the failure in the rollup log.

@Dylan-DPC 
@tmandry

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 98685a4bf2ef50c6d6a64ef3867a29994d5a4a25 has been approved by `tmandry`

<!-- @bors r=tmandry 98685a4bf2ef50c6d6a64ef3867a29994d5a4a25 -->
<!-- homu: {"type":"Approved","sha":"98685a4bf2ef50c6d6a64ef3867a29994d5a4a25","approver":"tmandry"} -->

richkadel commented:
Should fix #73386

Manishearth commented:
```
2020-06-16T06:21:09.6227910Z ---- [mir-opt] mir-opt/instrument_coverage.rs stdout ----
2020-06-16T06:21:09.6318830Z 
2020-06-16T06:21:09.6388710Z error: compilation failed!
2020-06-16T06:21:09.6422990Z status: exit code: 1
2020-06-16T06:21:09.6493760Z command: "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/stage2/bin/rustc" "/Users/runner/runners/2.170.1/work/1/s/src/test/mir-opt/instrument_coverage.rs" "-Zthreads=1" "--target=x86_64-apple-darwin" "-Zdump-mir=all" "-Zmir-opt-level=3" "-Zdump-mir-exclude-pass-number" "-Zdump-mir-dir=/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage" "-C" "prefer-dynamic" "--out-dir" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/native/rust-test-helpers" "-Zinstrument-coverage" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/auxiliary"
2020-06-16T06:21:09.6522630Z stdout:
2020-06-16T06:21:09.6560550Z ------------------------------------------
2020-06-16T06:21:09.6619440Z 
2020-06-16T06:21:09.6653760Z ------------------------------------------
2020-06-16T06:21:09.6694350Z stderr:
2020-06-16T06:21:09.6746420Z ------------------------------------------
2020-06-16T06:21:09.6783130Z error: linking with `cc` failed: exit code: 1
2020-06-16T06:21:09.6846280Z   |
2020-06-16T06:21:09.6933950Z   = note: "cc" "-m64" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/instrument_coverage.instrument_coverage.7rcbfp3g-cgu.0.rcgu.o" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/instrument_coverage.instrument_coverage.7rcbfp3g-cgu.1.rcgu.o" "-o" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/instrument_coverage" "-Wl,-dead_strip" "-no-pie" "-nodefaultlibs" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/native/rust-test-helpers" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/auxiliary" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" "-L" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" "-lstd-4ee1eeb3ff245df9" "/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-5ae1570bfb7fe005.rlib" "-lSystem" "-lresolv" "-lc" "-lm" "-Wl,-rpath,@loader_path/../../../stage2/lib/rustlib/x86_64-apple-darwin/lib" "-Wl,-rpath,/Users/runner/runners/2.170.1/work/1/s/lib/rustlib/x86_64-apple-darwin/lib"
2020-06-16T06:21:09.6966660Z   = note: clang: warning: argument unused during compilation: '-nopie' [-Wunused-command-line-argument]
2020-06-16T06:21:09.7028860Z           ld: warning: directory not found for option '-L/Users/runner/runners/2.170.1/work/1/s/build/x86_64-apple-darwin/test/mir-opt/instrument_coverage/auxiliary'
2020-06-16T06:21:09.7070920Z           Undefined symbols for architecture x86_64:
2020-06-16T06:21:09.7108460Z             "___llvm_profile_runtime", referenced from:
2020-06-16T06:21:09.7169740Z                 ___llvm_profile_runtime_user in instrument_coverage.instrument_coverage.7rcbfp3g-cgu.0.rcgu.o
2020-06-16T06:21:09.7233580Z                (maybe you meant: ___llvm_profile_runtime_user)
2020-06-16T06:21:09.7334980Z           ld: symbol(s) not found for architecture x86_64
2020-06-16T06:21:09.7410200Z           clang: error: linker command failed with exit code 1 (use -v to see invocation)
2020-06-16T06:21:09.7480460Z           
2020-06-16T06:21:09.7562460Z 
2020-06-16T06:21:09.7625570Z error: aborting due to previous error
```

From https://github.com/rust-lang/rust/pull/73388#issuecomment-644557159

@bors r-

richkadel commented:
IIUC, bors skips testing some targets by default, because the rollup failure occurred when testing on a Darwin target.

I normally build (locally) on Linux, but use a Mac laptop. I'm cloning my forked repo to test directly on the Mac, and try to repro the failure.

Q: Can I enable the macos/darwin target in the pre-merge bors tests in GitHub?

Also, I'm looking at attempting a fix.

Note that:
src/librustc_codegen_ssa/back/linker.rs:
```rust
    fn pgo_gen(&mut self) {
        if !self.sess.target.target.options.linker_is_gnu {
            return;
        }

        // If we're doing PGO generation stuff and on a GNU-like linker, use the
        // "-u" flag to properly pull in the profiler runtime bits.
        //
        // This is because LLVM otherwise won't add the needed initialization
        // for us on Linux (though the extra flag should be harmless if it
        // does).
        //
        // See https://reviews.llvm.org/D14033 and https://reviews.llvm.org/D14030.
        //
        // Though it may be worth to try to revert those changes upstream, since
        // the overhead of the initialization should be minor.
        self.cmd.arg("-u");
        self.cmd.arg("__llvm_profile_runtime");
    }
```

For the failing target:
src/librustc_target/spec/x86_64_apple_darwin.rs

It includes the Target setting:
```rust
        linker_flavor: LinkerFlavor::Gcc,
```

But it does not set:

```rust
        linker_is_gnu: true,
```

I may try adding this in my local build on Mac, and see if it resolves the linker issue.

`linker_is_gnu` is not a new setting, so I don't know if there is a reason it is not set for Darwin, but it is suspicious.

Does anyone have any insight?

Manishearth commented:
> Q: Can I enable the macos/darwin target in the pre-merge bors tests in GitHub?

@richkadel no, but you can ask people to use `try` to run the full CI suite without having to wait for the queue

richkadel commented:
> I may try adding this in my local build on Mac, and see if it resolves the linker issue.
> 
> `linker_is_gnu` is not a new setting, so I don't know if there is a reason it is not set for Darwin, but it is suspicious.

That didn't work, producing different errors due to linker flags that fail for Darwin, such as:

```
  = note: ld: unknown option: -O1
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

So there is a reason that Target option was not set.

I'll try some other possibilities.

wesleywiser commented:
> // See https://reviews.llvm.org/D14033 and https://reviews.llvm.org/D14030.

Those patches date back to 2015 so I wonder if our version of them has bit-rotted since then. Looking through current LLVM, it looks like this logic has been replaced with per-target overrides

https://github.com/llvm/llvm-project/blob/8422bc9efcbc1253281e2b8043a97d6e8f842234/clang/include/clang/Driver/ToolChain.h#L611-L614

with a similar Linux implementation

https://github.com/llvm/llvm-project/blob/ccd8b7b103470beb79140ecf9d6ccfaddf7fbc11/clang/lib/Driver/ToolChains/Linux.cpp#L875-L883

and a very different Darwin implementation

https://github.com/llvm/llvm-project/blob/6a3469f58d0c230e86043f6975f048968334dfa4/clang/lib/Driver/ToolChains/Darwin.cpp#L1159-L1202

All that to say, our implementation doesn't look correct with regards to upstream LLVM but I don't know enough to say for sure.

richkadel commented:
I managed to get the test to link and pass with this change (added check for target_os "macos"):

src/librustc_codegen_ssa/back/linker.rs:456
```rust
    fn pgo_gen(&mut self) {
        if !self.sess.target.target.options.linker_is_gnu
            && self.sess.target.target.target_os != "macos"
        {
            return;
        }
```

I'm not sure if there's a preference to fix this differently.

richkadel commented:
@wesleywiser - Are you able to set `try` to run the full CI suite with this change (including all targets)?

Manishearth commented:
@bors try

bors commented:
:hourglass: Trying commit bcdeac348e7c6a65f21c8427fb175f7396dae807 with merge 27b201f7642616f7b1d87bcb89be3e7c5eabbbad...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bcdeac348e7c6a65f21c8427fb175f7396dae807","merge_sha":"27b201f7642616f7b1d87bcb89be3e7c5eabbbad"} -->

mati865 commented:
To run tests on Mac with `bors try` you have to modify `src/ci/azure-pipelines/try.yml` by pasting there this code:
```
- job: macOS
  timeoutInMinutes: 600
  pool:
    vmImage: macos-10.15
  steps:
  - template: steps/run.yml
  strategy:
    matrix:
      x86_64-apple:
        SCRIPT: ./x.py test
        INITIAL_RUST_CONFIGURE_ARGS: --build=x86_64-apple-darwin --enable-sanitizers --enable-profiler --set rust.jemalloc
        RUSTC_RETRY_LINKER_ON_SEGFAULT: 1
        MACOSX_DEPLOYMENT_TARGET: 10.8
        MACOSX_STD_DEPLOYMENT_TARGET: 10.7
        NO_LLVM_ASSERTIONS: 1
        NO_DEBUG_ASSERTIONS: 1
```
Without that only Linux dist jobs are ran.

tmandry commented:
@bors try-

richkadel commented:
@mati865 How do I try for *all* targets tested by bors before a merge?
I'm pretty sure my fix addresses the last failure, but I want to confirm there are no others.

tmandry commented:
Enabled mac and windows.

@bors try

bors commented:
:hourglass: Trying commit c2de7803887cb541195ab13ced5185b572caf290 with merge 9c5276ea9bf5ee4bfd77013010a7fc75431b1775...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c2de7803887cb541195ab13ced5185b572caf290","merge_sha":"9c5276ea9bf5ee4bfd77013010a7fc75431b1775"} -->

tmandry commented:
> `try-` doesn't work btw :P

As in it's broken, or our version of bors doesn't match what's documented [here](https://bors.tech/documentation/)? I've always been confused by what does and doesn't support.

mati865 commented:
@tmandry you have enabled only dist builds. They do **not** run tests.

@richkadel due to CI capacity you cannot enable all tests at same time for try builds. They will be ran when PR is approved.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32156&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32156/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32156/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 4'
Agent machine name: 'fv-az619'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200604.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200604.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/2c3eb3f1-813e-428e-8453-3f9509d4fe5c.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73011/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73011/merge:refs/remotes/pull/73011/merge
---
 ---> f883e675ad62
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> c0b156eb069c
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 8541bab6b38c
Successfully built 8541bab6b38c
Successfully tagged rust-ci:latest
Built container sha256:8541bab6b38c07f1b7eb787539b9cbe93daa6ac4458d3d7bd8a8921622a14ba1
---
  local time: Tue Jun 16 18:53:12 UTC 2020
  network time: Tue, 16 Jun 2020 18:53:12 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73011/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73011/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (8108) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Dylan-DPC commented:
> As in it's broken, or our version of bors doesn't match what's documented here? I've always been confused by what bors does and doesn't support.

i'm not sure, will have to check with infra. But i checked the queue and the try is still running

Dylan-DPC commented:
> As in it's broken, or our version of bors doesn't match what's documented here? I've always been confused by what bors does and doesn't support.

i'm not sure, will have to check with infra. But i checked the queue and the try is still running

tmandry commented:
Okay, trying again. With tests enabled for a mac and windows target this time, hopefully.

@bors try

bors commented:
:hourglass: Trying commit d88ec6a9b0357484c35ca341398135b54d4b4682 with merge 7e95261094f3218518acf2987e4173092b8c7b5b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d88ec6a9b0357484c35ca341398135b54d4b4682","merge_sha":"7e95261094f3218518acf2987e4173092b8c7b5b"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32161)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32161","builder_name":"checks-azure"} -->

Manishearth commented:
@tmandry try- dequeues a job, but if something is running the tests will continue to run, bors isn't able to cancel checks

richkadel commented:
@wesleywiser - FYI, the change I had suggested here:

> src/librustc_codegen_ssa/back/linker.rs:456
> ```rust
>     fn pgo_gen(&mut self) {
>         if !self.sess.target.target.options.linker_is_gnu
>             && self.sess.target.target.target_os != "macos"
>         {
>             return;
>         }
> ```

... seems to have been a false lead. I can see the mir-opt tests passed locally on my Mac after this change, but I can't reproduce it, and in fact, I don't see how this would have ever worked. (Mystery)

I think I have a reproducible, working fix. I'm validating on a clean build before I report more details. I'll update the thread shortly.

richkadel commented:
OK, I was able to reproduce the bors queue build error on Darwin. (Windows build errors were the same, so hopefully the fixes included in my last patch address both of them.)

1. Added a comment to `config.toml.example` to highlight the requirement to enable `profiler = true` when compiling with `-Z instrument-coverage` (and other profile flags). This isn't obvious, so hopefully the comment will be helpful.
1. Added a check for the `instrument-coverage` option to load the profiler runtime. Otherwise, it was only loaded for the profile options.
1. Following the same pattern from existing tests, I added `needs-profiler-support` to the test configuration for mir-opt/instrument_coverage.
1. Updated test diffs affected by the line number changes after inserting `needs-profiler-support`.

Note the previous change (checking for "macos" in `pgo_gen()`) was unnecessary, so I removed that change.

I decided not to reenable the `try` since I now understand, and believe I've fixed the errors, in a platform-independent way.

@tmandry can you re-approve so we are waiting-on-bors again?

Thanks!

bjorn3 commented on src/libcore/intrinsics.rs:
```diff
@@ -1941,6 +1941,13 @@ extern "rust-intrinsic" {
+    /// generation.
+    #[cfg(not(bootstrap))]
+    #[lang = "count_code_region"]
+    pub fn count_code_region(_index: u32);
```
```suggestion
    pub fn count_code_region(index: u32);
```

No need to use `_` here.

richkadel replied:
Right... Cut-and-paste error. Thanks!

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit c3387293d4ff050cce2c7f3a79c8e7b040515f8c has been approved by `tmandry`

<!-- @bors r=tmandry c3387293d4ff050cce2c7f3a79c8e7b040515f8c -->
<!-- homu: {"type":"Approved","sha":"c3387293d4ff050cce2c7f3a79c8e7b040515f8c","approver":"tmandry"} -->

richkadel commented:
@Dylan-DPC - Would you be willing to include this in the next rollup? Thanks!

richkadel commented:
> @Dylan-DPC - Would you be willing to include this in the next rollup? Thanks!

Please hold off. We're going to do another bors try with macos and windows tests enabled, to make sure there are no more surprises.

tmandry commented:
@bors try

bors commented:
:hourglass: Trying commit b9f0304af8b74bdf01cd39f1603d73bdc599de79 with merge 7e5ef77edd62e7082b227400416e103834f1ef1a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b9f0304af8b74bdf01cd39f1603d73bdc599de79","merge_sha":"7e5ef77edd62e7082b227400416e103834f1ef1a"} -->

mati865 commented:
This PR is next in bors queue. It'd be safer for rollups to not include it.

Dylan-DPC commented:
it's not next on the queue :thinking: it's on "try" which is separate so it won't be merged

mati865 commented:
@Dylan-DPC right, I mistook `73011` for `70311` :open_mouth:

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32224)
Build commit: 7e5ef77edd62e7082b227400416e103834f1ef1a (`7e5ef77edd62e7082b227400416e103834f1ef1a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32224"},"merge_sha":"7e5ef77edd62e7082b227400416e103834f1ef1a"} -->

richkadel commented:
I removed the "try" config changes (PR passes on all platforms now!).

This version should be good to merge, assuming no new merge conflicts.

Thanks!

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 36c9014ddd3e2ac6b6a0e9f623e791281c40473d has been approved by `tmandry`

<!-- @bors r=tmandry 36c9014ddd3e2ac6b6a0e9f623e791281c40473d -->
<!-- homu: {"type":"Approved","sha":"36c9014ddd3e2ac6b6a0e9f623e791281c40473d","approver":"tmandry"} -->

RalfJung commented on src/librustc_mir/interpret/intrinsics.rs:
```diff
@@ -389,6 +389,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 self.copy_op(self.operand_index(args[0], index)?, dest)?;
             }
+            // FIXME(#73156): Handle source code coverage in const eval
+            sym::count_code_region => (),
```
Please always ping @rust-lang/wg-const-eval for new intrinsics in CTFE.
It is not great when I realize by mere accident that we have a new intrinsic here.

RalfJung replied:
Also, note that as per this comment, T-lang should have been involved as well:

https://github.com/rust-lang/rust/blob/394e1b40d264aa6928811919c1124fa248e7d802/src/libcore/intrinsics.rs#L8-L9

oli-obk replied:
amusingly this intrinsic is *not* marked as const at all. The static checks preventing non-const intrinsics from being invoked happen before mir optimizations. Since instrumentation is a MIR transformation... this isn't caught. Maybe instrumentation should happen before const checking and such?

RalfJung replied:
Oh, so this intrinsic is not called by users, but the call is inserted by an instrumentation pass later? Interesting. That reduces my T-lang concern.

> Maybe instrumentation should happen before const checking and such?

Maybe -- I am not sure to what extend it would be useful to check the instrumented code as if it was user-written.

bjorn3 replied:
If I write

```rust
const fn a() -> u8 {
    42
}

const A: u8 = a();

fn main() {
    println!("{}", A);
}
```

Then currently `a` will contribute towards coverage, yet it is not marked as executed. I think the `count_code_region` intrinsic should also mark code as executed during CTFE.

RalfJung replied:
@bjorn3 that is a separate issue though from the other discussion here? Namely, it is https://github.com/rust-lang/rust/issues/73156.

bjorn3 replied:
Didn't see that issue, thanks.

wesleywiser replied:
Sorry about that @RalfJung! That was my fault; I didn't realize we should be pinging that group. 

Do you think we still need to get the lang team involved? This intrinsic is used by an unstable `-Z` compiler flag which generates code coverage data using LLVM features. The compiler team already signed off on this approach via an MCP.

RalfJung replied:
I doubt they will have any objections, but sending T-lang an "FYI this is what we did here" would still be a good idea I think.

richkadel replied:
Thanks for flagging the issue @RalfJung ...

Note also, the next PR increment is under review, and adds some additional Rust intrinsics that I believe we need for coverage map generation.

See this related note:

https://github.com/rust-lang/rust/pull/73684#discussion_r446559720

Since the MIR analysis is minimal at the moment (injecting at the function level only) I'm not generating calls to these intrinsics yet, but when I do, I suspect I may need to add these new intrinsics to src/librustc_mir/interpret/intrinsics.rs as well, similarly stubbed out.
